model/DEL_network.py

```diff
@@ -0,0 +1,278 @@
+        cholesky_raw = q
+        for layer in self.model_layers:
+            cholesky_raw = layer(cholesky_raw)
+
+        diag = cholesky_raw[0:self.num_states-1] + diag_bias
+        lower_tri = cholesky_raw[self.num_states-1:]
+
+        cholesky = torch.diag(diag)
+        lower_tri_indices = np.tril_indices(self.num_states-1, -1)
+        cholesky[lower_tri_indices] = lower_tri
+        M = cholesky @ cholesky.T
+
+        return M
+
+
+class PotentialEnergyNetwork(torch.nn.Module):
+    def __init__(self, q_size, hidden_list):
+        """
+        Neural Network used to approximate a lagrangian systems potential energy
+        """
+        super(PotentialEnergyNetwork, self).__init__()
+        self.hyperparams = None
+        self.model_layers = torch.nn.ModuleList()
+
+        # input layer - Lagrange
+        self.model_layers.append(torch.nn.Linear(q_size, hidden_list[0]))
+        self.model_layers.append(torch.nn.Dropout())
+        self.model_layers.append(torch.nn.Tanh())
+        # add all hidden layers
+        for i in range(0, len(hidden_list) - 1):
+            self.model_layers.append(torch.nn.Linear(hidden_list[i], hidden_list[i + 1]))
+            self.model_layers.append(torch.nn.Dropout())
+            self.model_layers.append(torch.nn.Tanh())
+
+        # output layer
+        # output is always one from this network because it is calculating system energy
+        self.model_layers.append(torch.nn.Linear(hidden_list[-1], 1))
+        self.model_layers.append(torch.nn.Dropout())
+        self.model_layers.append(torch.nn.Softplus())
+
+    def forward(self, q):
+        for layer in self.model_layers:
+            q = layer(q)
+
+        return q
+
+
+class DissipativeForceNetwork(torch.nn.Module):
+    def __init__(self, D_in, hidden_list):
+        """
+        Neural Network used to the represent dissipative forces such as drag, friction, etc
+        """
+        super(DissipativeForceNetwork, self).__init__()
+        self.hyperparams = None
+        self.model_layers = torch.nn.ModuleList()
+
+        # input layer - Lagrange
+        self.model_layers.append(torch.nn.Linear(D_in, hidden_list[0]))
+        self.model_layers.append(torch.nn.Dropout())
+        self.model_layers.append(torch.nn.Tanh())
+        # add all hidden layers
+        for i in range(0, len(hidden_list) - 1):
+            self.model_layers.append(torch.nn.Linear(hidden_list[i], hidden_list[i + 1]))
+            self.model_layers.append(torch.nn.Dropout())
+            self.model_layers.append(torch.nn.Tanh())
+
+        # output layer
+        # output is always one from this network because it is calculating system energy
+        self.model_layers.append(torch.nn.Dropout())
+        self.model_layers.append(torch.nn.Linear(hidden_list[-1], 6))
+
+    def forward(self, q, q_next, h):
+        r1 = q[0:3]
+        Q1 = q[3:7]
+        r2 = q_next[0:3]
+        Q2 = q_next[3:7]
+        omega = 2 / h * (H.T @ L(Q1).T @ Q2)
```

Comment:
This seems like the best approach, but I suppose we could use the omega in the state here if we wanted?